limits/aio/storage/memcached.py

```diff
@@ -150,3 +229,91 @@ async def check(self) -> bool:
 
     async def reset(self) -> Optional[int]:
         raise NotImplementedError
+
+    def _current_window_key(
+        self, key: str, expiry: int, now: Optional[float] = None
+    ) -> str:
+        if now is None:
+            now = time.time()
+        return f"{key}/{int(now / expiry)}"
+
+    def _previous_window_key(
+        self, key: str, expiry: int, now: Optional[float] = None
+    ) -> str:
+        if now is None:
+            now = time.time()
+        return f"{key}/{int((now - expiry) / expiry)}"
+
+    async def acquire_sliding_window_entry(
+        self,
+        key: str,
+        limit: int,
+        expiry: int,
+        amount: int = 1,
+    ) -> bool:
+        now = time.time()
+        previous_key = self._previous_window_key(key, expiry, now)
+        current_key = self._current_window_key(key, expiry, now)
+
+        if amount > limit:
+            return False
+
+        (
+            previous_count,
+            previous_ttl,
+            current_count,
+            _,
+        ) = await self._get_sliding_window_info(
+            previous_key, current_key, expiry, get_current_ttl=False
+        )
+
+        weighted_count = previous_count * previous_ttl / expiry + current_count
+        if weighted_count + amount > limit:
+            return False
+        else:
+            # Hit, increase the current counter.
+            # If the counter doesn't exist yet, set twice the theorical expiry.
+            await self.incr(current_key, expiry + ceil(now % expiry), amount=amount)
```

Comment:
I think changing this to use `expiry * 2` would ensure the window stats are accurate and the failing test case is addressed.

Reply:
You're right. The delay should be set to expiry * 2, otherwise the counter might be reset too soon. I modified the sync implementation, I have to do the same with the async one.
However, the current failing test is not related to this. I think it's because the Memcached timer behavior is different from the others. Since I use a timestamp-based key, the reset time depends on the local machine timestamp (like the fixed window algorithm), whereas with Redis, it depends on the time of the first hit (like the moving window algorithm). I'm working on it.

Reply:
Thanks for sharing your findings! I've been trying to wrap my head around some of the edge cases (while working on the mongodb/etcd storage implementations). 

At the moment, I'm still seeing a few (flaky) test results:
- [mongodb](https://github.com/alisaifee/limits/tree/sliding-window-counter-mongodb) I've tried to emulate your implementation in redis 
- [etcd](https://github.com/alisaifee/limits/tree/sliding-window-etcd) Similar to your memcached implementation but without actually decrementing in case of a race condition allowing an increment to exceed the current window

Reply:
I did some more tests. We could allow exceeds in the current window. However, I cannot guarantee the upper limit of excess hits when the backend storage is under pressure. When testing locally, I did have 3-4 more hits than expected. Then I tried to add random delay on the `get` and `incr` operations (eg: `await asyncio.sleep(random.random() / 1000)`, and I could get not one, but more than a thousand more hits than expected. That's why I added the decrementation.

I didn't find a way to simulate a read/write delay properly in my tests.

Reply:
With the etcd & mongodb implementions, my naive solution was to fail the acquisition if the increment on the current window resulted in exceeding the limit and that passed the concurrency tests - however, both mongodb and etcd are not at all comparable in terms of performance with redis & memcached.